### PR TITLE
Discovery service

### PR DESCRIPTION
**Pull request by naveed050**
added code to create Instance from AWSInstance for discoveryService route
added  custom write for jsonFormat of Instance as it has more than 22 fields. did not write custom read

### CODE REVIEW

File: src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala [hunk 1]
```diff
@@ -8,28 +9,30 @@ import org.slf4j.LoggerFactory
   * Created by sampathr on 22/9/16.
   */
 case class ImageInfo(override val id: Option[Long],
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala:11]: Even the id is optional?

File: src/main/scala/com/imaginea/Main.scala [hunk 2]
```diff
@@ -59,18 +56,148 @@ object Main extends App {
...
+      }
+    }
+  }
+  implicit val ipPermissionInfoFormat = jsonFormat6(IpPermissionInfo.apply)
+  implicit val securityGroupsFormat = jsonFormat7(SecurityGroupInfo.apply)
+  implicit object InstanceFormat extends RootJsonFormat[Instance] {
+
+    def write(i: Instance): JsValue = {
+
+      val fieldName1 = "id"
```
Comment (eklavya) [inline src/main/scala/com/imaginea/Main.scala]: This could definitely be more terse. Why are we doing it in the max lines possible way?

**Review by eklavya — CHANGES_REQUESTED** [review]

File: src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala (hunk 1 above)
Comment (naveed050) [reply src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala:11]: id is optional in all case classes as we assign the value to it  after getting the id by saving it to graph db

File: src/main/scala/com/imaginea/Main.scala (hunk 2 above)
Comment (naveed050) [reply src/main/scala/com/imaginea/Main.scala]: I will look into it

File: src/main/scala/com/imaginea/Main.scala [hunk 3]
```diff
@@ -59,15 +56,127 @@ object Main extends App {
...
+  implicit object InstanceFormat extends RootJsonFormat[Instance] {
+
+    def write(i: Instance): JsValue = {
+
+      val fieldNames = List("id", "instanceId", "name", "state", "instanceType", "platform", "architecture", "publicDnsName", "launchTime", "memoryInfo", "rootDiskInfo",
+        "tags", "sshAccessInfo", "liveConnections", "estimatedConnections", "processes", "image", "existingUsers", "account", "availabilityZone", "privateDnsName",
+        "privateIpAddress", "publicIpAddress", "elasticIp", "monitoring", "rootDeviceType", "blockDeviceMappings", "securityGroups", "reservedInstance", "region")
+
+      val fields = new collection.mutable.ListBuffer[(String, JsValue)]
+      fields ++= longToJsField(fieldNames(1), i.id)
```
Comment (eklavya) [inline src/main/scala/com/imaginea/Main.scala:105]: Why can't you just do:

``` scala
JsObject(
  "id" -> i.id,
  ...
)
```

File: src/main/scala/com/imaginea/activegrid/core/models/AWSComputeAPI.scala [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+      val privateIpAddress = Option(instance.getPrivateIpAddress)
+      val publicIpAddress = Option(instance.getPublicIpAddress)
+      val instanceType = Option(instance.getInstanceType)
+      val availabilityZone = if (instance.getPlacement != null) Option(instance.getPlacement.getAvailabilityZone) else None
+      val monitoring = if (instance.getMonitoring != null) Option(instance.getMonitoring.getState) else None
+      val rootDeviceType = Option(instance.getRootDeviceType)
+      val nodeName = publicDnsName.get
+      val memoryInfo = Some(StorageInfo(None, 0D, AWSInstanceType.toAWSInstanceType(instance.getInstanceType).ramSize))
+      val rootDiskInfo = Some(StorageInfo(None, 0D, AWSInstanceType.toAWSInstanceType(instance.getInstanceType).rootPartitionSize))
+      import scala.collection.JavaConversions._
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/models/AWSComputeAPI.scala]: Why not import at the top?

File: src/main/scala/com/imaginea/Main.scala (hunk 3 above)
Comment (naveed050) [reply src/main/scala/com/imaginea/Main.scala:105]: id is optional (along with other fields)  , so we need to check first if there is some value or not. Only then we are adding it to JsObject

File: src/main/scala/com/imaginea/activegrid/core/models/AWSComputeAPI.scala (hunk 4 above)
Comment (naveed050) [reply src/main/scala/com/imaginea/activegrid/core/models/AWSComputeAPI.scala]: Yes Saurabh, removed it now

**Pull request merged**